File: app/api/entities/story.rb
class Entities::Story < Entities::BaseEntity
  expose :id
  expose :title
  expose :description
  expose :story_type
  expose :estimate
  expose :state
  expose :labels
  expose :requested_by_name
  expose :cycle_time
  expose :owned_by_name
  expose :owned_by_initials
  expose :accepted_by_name

  with_options(format_with: :iso_timestamp) do
    expose :created_at, if: ->(story, _) { story.created_at.present? }
    expose :started_at, if: ->(story, _) { story.started_at.present? }
    expose :accepted_at, if: ->(story, _) { story.accepted_at.present? }
  end
end

File: app/assets/javascripts/components/shapes/story.js
import PropTypes from "prop-types";
import NotePropTypes from "./note";
import TaskPropType from "./task";

export const storyPropTypes = {
  id: PropTypes.oneOfType([PropTypes.number, PropTypes.symbol]),
  title: PropTypes.string,
  description: PropTypes.string,
  estimate: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  storyType: PropTypes.string,
  state: PropTypes.string,
  acceptedAt: PropTypes.string,
  requestedById: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  ownedById: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  projectId: PropTypes.number,
  createdAt: PropTypes.string,
  updatedAt: PropTypes.string,
  position: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  labels: PropTypes.array,
  requestedByName: PropTypes.string,
  ownedByName: PropTypes.string,
  ownedByInitials: PropTypes.string,
  releaseDate: PropTypes.string,
  deliveredAt: PropTypes.string,
  errors: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  notes: PropTypes.arrayOf(NotePropTypes.isRequired),
  tasks: PropTypes.arrayOf(TaskPropType.isRequired),
};

const storyPropTypesShape = PropTypes.shape(storyPropTypes).isRequired;

export default storyPropTypesShape;
